### PR TITLE
[match] using ruby openssl instead of terminal one

### PR DESCRIPTION
**Pull request by Cyberbeni**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
https://github.com/fastlane/fastlane/issues/10306
https://github.com/fastlane/fastlane/issues/9542

My current solution does exactly the same as the master on macOS 10.12.6

We should also change any other terminal based openssl usage to the ruby version to avoid these kind of conflicts.

We should also deprecate the current encryption of certificates to be more safer:
-randomly generate an IV instead of calculating it from the password and provide it in the certs repo unencrypted
-use higher iteration count for generating the key
https://crypto.stackexchange.com/questions/2280/why-is-the-iv-passed-in-the-clear-when-it-can-be-easily-encrypted
https://stackoverflow.com/questions/1949640/does-iv-work-like-salt
https://ruby.github.io/openssl/OpenSSL/Cipher.html

### Description
Changed terminal openssl to ruby openssl in fastlane match

### CODE REVIEW

**Comment by Cyberbeni** [comment]
I signed it!

**Comment by KrauseFx** [comment]
Awesome, thanks, same question as https://github.com/fastlane/fastlane/pull/11652#issuecomment-359449856

**Comment by Cyberbeni** [comment]
Yes, there are also tests that fail if it wouldn't work correctly.

**Comment by KrauseFx** [comment]
Tests are great, but they don't cover everything 😉

**Comment by Cyberbeni** [comment]
Yeah, but I also tested this one thoroughly because at first I had some problems figuring out what to use to get the exact same result as the openssl that's preinstalled on macOS 10.12.6

File: match/lib/match/encrypt.rb
Comment (lacostej) [inline match/lib/match/encrypt.rb]: * any reason we catch errors only in decrypt and not in encrypt?
* why logging the error only in verbose?

File: match/lib/match/encrypt.rb
Comment (lacostej) [inline match/lib/match/encrypt.rb]: if we start depending on a new gem, we should modify the Gemfile. If we want to enforce a way to use it, maybe we should provide a wrapper.

File: match/lib/match/encrypt.rb
Comment (lacostej) [inline match/lib/match/encrypt.rb]: shorter with `data_to_encrypt = File.read(path)`

File: match/lib/match/encrypt.rb
Comment (lacostej) [inline match/lib/match/encrypt.rb]: shorter with `File.write..`.

File: match/lib/match/encrypt.rb
Comment (Cyberbeni) [reply match/lib/match/encrypt.rb]: Which new gem? OpenSSL? It's already used in most places where we want to work with SSL related stuff and we want to use the latest version possible, so we have the latest security updates.
We don't want to enforce the way it's used in this current pull request because that's deprecated, it's just the way how the command line openssl command worked.
After this PR is merged, we can work out a better way of encrypting stuff (generate a random IV instead of generating it from the password, like the key, and provide it unencrypted. Reasons why this is more secure is explained in the links in the PR) and do all new encrypting that way, while still supporting decrypting with the old algorithm if the users' match repo haven't been encrypted in the new way yet.
The main goal for this PR is to solve the error for people who installed OpenSSL 1.1.x on their machine (and work with people who didn't) because that changes the default hashing algorithm and it wasn't implicitly told to the old openssl command to use MD5.

**Comment by Cyberbeni** [comment]
Okay, I really should try out RubyMine instead of Atom while I still have my all product license going. Feels like some of my newbie mistakes could've made an instant warning instead of having to run the tests.

**Comment by janpio** [comment]
@Cyberbeni Could you please create an account at appveyor.com, add your fork of `fastlane` as a project and invite use "janpio" to it? I am in the progres of adding Windows CI to fastlane, but couldn't find a way to run the tests on your fork+branch from my side :/

(If you don't want to do this, we can also wait until this is merged and see if it breaks stuff on Windows CI for fastlane and **then** fix it here with another PR)

**Comment by Cyberbeni** [comment]
@janpio It asks for e-mail when I want to add collaborator for the team.

**Comment by janpio** [comment]
piotrowski+github@gmail.com should work.

**Comment by janpio** [comment]
9 failures, but unrelated to your changes. This is good to merge from a Windows viewpoint then.

```
Finished in 2 minutes 34.3 seconds (files took 3.7 seconds to load)
5213 examples, 9 failures, 180 pending
Failed examples:
rspec ./fastlane/spec/actions_specs/automatic_code_signing_spec.rb:10 # Fastlane Automatic Code Signing enable_automatic_code_signing
rspec ./fastlane/spec/actions_specs/automatic_code_signing_spec.rb:33 # Fastlane Automatic Code Signing disable_automatic_code_signing for specific target
rspec ./fastlane/spec/actions_specs/automatic_code_signing_spec.rb:55 # Fastlane Automatic Code Signing disable_automatic_code_signing
rspec ./fastlane/spec/actions_specs/automatic_code_signing_spec.rb:75 # Fastlane Automatic Code Signing sets team id
rspec ./fastlane/spec/actions_specs/automatic_code_signing_spec.rb:102 # Fastlane Automatic Code Signing sets code sign identity
rspec ./fastlane/spec/actions_specs/automatic_code_signing_spec.rb:134 # Fastlane Automatic Code Signing sets profile name
rspec ./fastlane/spec/actions_specs/automatic_code_signing_spec.rb:166 # Fastlane Automatic Code Signing sets profile uuid
rspec ./fastlane/spec/actions_specs/automatic_code_signing_spec.rb:198 # Fastlane Automatic Code Signing sets bundle identifier
rspec ./fastlane/spec/actions_specs/automatic_code_signing_spec.rb:230 # Fastlane Automatic Code Signing targets not found notice
```

**Comment by janpio** [comment]
I really like how you are removing two really ugly hacks I had to use here by the way :)

File: match/lib/match/encrypt.rb
Comment (janpio) [reply match/lib/match/encrypt.rb]: @lacostej Any work necessary here or is this good?

**Comment by janpio** [comment]
Note to myself:
Also merge https://github.com/fastlane/fastlane/pull/11720 and https://github.com/fastlane/fastlane/pull/11652 when this gets merged.

File: match/lib/match/encrypt.rb
Comment (lacostej) [reply match/lib/match/encrypt.rb]: @Cyberbeni you're right. It's not new. It's just not yet depended upon in the Gemfile. Weird.

Question: are we 100% sure no one managed to encrypt data with openssl 1.1 and its new hashing algorithm and thus wouldn't be able to decrypt it with the previous hash? If so, shouldn't we use a decrypt mechanism that tries with both hash methods?

Comment (Cyberbeni) [reply match/lib/match/encrypt.rb]: @lacostej it doesn't hurt to fallback to SHA256 if decrypting with MD5 fails.
could only edit in browser, hopefully no errors.

Comment (lacostej) [reply match/lib/match/encrypt.rb]: @Cyberbeni it looks OK. I would probably add a few notes in the code

* that we use MD5 as primary for encryption even if known to not be the latest/most secure and that this whole encrypt/decrypt algorithm is meant to be improved
* that we fallback to decryption using SHA256 in case people encrypted data using openssl 1.1.0c or later with fastlane <= 2.80.0

or something like that.

Last question: given that the `pkcs5_keyivgen` method is deprecated (https://apidock.com/ruby/v1_9_3_125/OpenSSL/Cipher/pkcs5_keyivgen), was there a non deprecated way to achieve the same results?

Comment (Cyberbeni) [reply match/lib/match/encrypt.rb]: @lacostej They recommend using OpenSSL::KDF::pbkdf2_hmac instead which has almost the same parameters and only 1 output. pkcs5_keyivgen generates both the key(16 length) and IV(32 length), I tried with same inputs and 16 length and it resulted in a different key, so my guess is that there is no other, non-deprecated method that keeps us backwards compatible with older fastlane versions.

I'll add comments then.

**Review by lacostej — APPROVED** [review]
Thanks a lot @Cyberbeni for taking time to explain all the inner workings and trying various approaches.

**Comment by Cyberbeni** [comment]
@lacostej no problem. A coworker and me spent a good hour or two trying to find out why match didn't work on my MacBook, just wanted to help others with OpenSSL 1.1 avoid that.

**Comment by janpio** [comment]
🥇

**Pull request merged**
